### DISCUSSION TITLE
Adding a filter text input to a dropdown menu

### DISCUSSION THREAD

**Discussion by DaniGuardiola** [radix-ui/primitives#787] [Help]
I'm trying to create a dropdown menu that contains a text input at the top to filter among the options. The main problem I've faced is that the typeahead feature is conflicting with it, and removes focus from the text input to give it to the corresponding menu item.

For example, if the options are "America", "Europe" and "Antarctica" and I start typing "Europe" in the input, as soon as I type the first "E" it will switch focus over to the "Europe" item.

This is my main problem for now, although I do want to implement a few other behaviors, like:

- When the input is focused, on arrow down, focus the topmost menu item.
- When the topmost menu item is focused, on arrow up, focus the input.
- When input is focused, on ESC key down, close the menu.
- When input is NOT focused, but some menu item is, and the user presses a "text key" (e.g. not return or ESC), focus the input and populate it with the typed text.
- When input is focused, has a value and there's at least one option after filtering, highlight the topmost option. If return is pressed, trigger the onSelect handler for that option and close the menu (similar to just focusing that option directly and pressing return).
- When input is focused and tab / shift+tab is pressed, focus a menu item (this doesn't work on first render for some reason, maybe has something to do with the roving focus?)
- When a menu item is focused and tab / shift+tab is pressed, focus the text editor.

As I said, my main issue is the typeahead feature conflicting with the input. I've tried a bunch of things like preventing default, etc, but although I found some hacky ways to make it work, they all seem very fragile. How could I go about this?

My wishlist, in order of preference is:

- Add search/filtering support to the primitive natively. Ideally with support for your own text input component (like with slots or something like that).
- Allow disabling typeahead support.
- Improved event handling customization, with more escape hatches.

Thoughts? Thanks in advance :)

Update: I sort of fixed it by setting `textValue` to `""` (empty string) on all items. Still feels pretty hacky though, just sharing in case it's useful to anyone.

**Comment by jjenzz** [comment]
Hi @DaniGuardiola 👋 

It sounds like you are trying to convert the `DropdownMenu` into a Combobox but the latter is a [very different pattern in terms of accessibility](https://www.w3.org/TR/wai-aria-practices-1.1/#combobox). It's unlikely that your component will be accessible with your current approach as the `DropdownMenu` is not intended for that use-case, it follows the [Menu Button pattern](https://www.w3.org/TR/wai-aria-practices/#menubutton).

We do have plans to build a Combobox, but it's not in progress at the moment I'm afraid 😞

**Reply by DaniGuardiola** [reply]
Thanks for you response Jenna! That's what I was worried about, although I'm not entirely sure that pattern (combobox) fits my use case either. Everything behaves like a menu button:

- it is fundamentally stateless, just a bunch of options with event handlers, unlike select or combobox.
- triggered by a button
- same keyboard interactions for the most part (arrows, ESC, etc)
- pretty much everything else too

The only difference really is that, because of the potentially elevated number of options in my use case, we need a search input that filters the options down to a few. In that regard, it is similar to a combobox, but it's still stateless.

Do you still think this is a combobox? I admit I really need to get more familiar with a11y stuff, but I haven't been able to find the right pattern for this, so I just thought I'd leverage the dropdown menu component and hack a text input into it. My goal is to provide good a11y still, but I'm not sure how to achieve this.

In any case, I don't think I will be able to come up with a different approach (e.g. implementing a combobox) because of our project constraints. Do you think it'd be possible to improve the "escape hatches" or customization of the dropdown menu component? For instance, by adding a `disableTypeahead` prop of sorts, or `onTypeaheadKeyDown` where we could do `preventDefault`, as with other events.

Thank you!

**Reply by jjenzz** [reply]
Oooh, so you have something like this?

![image](https://user-images.githubusercontent.com/175330/126177500-7a69830d-72cd-45c4-9d2e-939340f1f867.png)

You're right, I wouldn't think that is a Combobox either 🤔 

> As I said, my main issue is the typeahead feature conflicting with the input. I've tried a bunch of things like preventing default, etc, but although I found some hacky ways to make it work, they all seem very fragile. How could I go about this?

It sounds like `event.stopPropagation()` in the `input`s `onKeyDown` event is what you need here. 

Separately, usually for this type of control your focus wouldn't leave the input and you would [manage focus using `aria-activedescendant`](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_focus_activedescendant). It wouldn't really be very accessible to keep moving someones focus as they type like you describe here:

> - When input is NOT focused, but some menu item is, and the user presses a "text key" (e.g. not return or ESC), focus the input and populate it with the typed text.
> - When input is focused, has a value and there's at least one option after filtering, highlight the topmost option. 

I can't be certain you won't run into issues trying to apply the `aria-activedescendant` pattern to the `DropdownMenu` though as we're already applying the [roving tabindex pattern](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_roving_tabindex). I'm not sure of the accessibility implications of combining the two but might be worth a shot 😬

**Reply by DaniGuardiola** [reply]
Hi Jenna, thanks for your response! Yes, that is accurate :)

> It sounds like event.stopPropagation() in the inputs onKeyDown event is what you need here.

I tried that among other things, it didn't work :/ I just tried again, both on the React event and the native one, and still no dice.

> Separately, usually for this type of control your focus wouldn't leave the input and you would manage focus using aria-activedescendant. It wouldn't really be very accessible to keep moving someones focus as they type like you describe here: (...)

Well, the focus here would really only change when you start typing. While you type, the focus would remain on the input, but the topmost option would be highlighted, and if you press return it would be "selected". I can see the resemblance to the aria-activedescendant behavior here, but I'm not sure that it would work properly with the roving tabindex, as you mention. I'll keep working on this tomorrow.

Any chance this could be supported by Radix natively? `DropdownMenu.SearchInput`? As an alternative to typeahead? :P

In fact I'd be down to giving it at try myself and sending a PR. Been looking for an excuse to contribute to this awesome project. Let me know if this is something that'd make sense to implement.

**Reply by jjenzz** [reply]
> Let me know if this is something that'd make sense to implement.

Tbh, I think this would be a separate component if we were to provide it. It may look similar to a `DropdownMenu` but it has its own separate accessibility concerns. 

> I tried that among other things, it didn't work :/

It seems to work for me here https://codesandbox.io/s/recursing-solomon-bxvmy. You should be able to type in the input there without anything stealing focus.

**Reply by DaniGuardiola** [reply]
Thanks, will keep that in mind. Let me know if y'all would be open to the possibility of adding such component. I know it's not a very common pattern. And thanks a lot for the sandbox too, although for some very weird reason, that's not working on my actual implementation :/ I'll have to do some digging.

**Reply by dcastil** [reply]
I also like the UX of a "select with filter input", especially for these reasons:
1. When there are a lot of options, it's easier to filter the list down and reduce the number of options rather than just navigating closer to the option I want in the full list.
2. I don't have to search for the selected option with my eyes but instead can keep looking at the first option in the list as I type.
3. It's easier to discover for new users that they can use the keyboard to navigate through the options when there is an input present.
4. It's faster to navigate through filtered list, especially when options are not sorted alphabetically or fuzzy search is used.

I've implemented such components myself and I see more companies using this pattern. E.g. [Linear](https://linear.app) is relying on this pattern for most of their selects:
<img width="331" alt="Screenshot 2022-04-08 at 10 46 20" src="https://user-images.githubusercontent.com/31006608/162400345-85a76c54-31ba-4b5c-8fba-7d67d620b023.png">

However, I don't see this pattern in any UI components library like Radix or Headless UI. I'd love to see this component in Radix or to get the option to use a filter input instead of the typeahead for the select component.

**Reply by benoitgrelard** [reply]
Hey @dcastil, in the case of Linear, it's basically a ComboBox which we is on our list of primitives to tackle.

**Reply by dcastil** [reply]
@benoitgrelard that's awesome! But isn't a combobox triggered by an input instead of a button?

**Reply by benoitgrelard** [reply]
Oh that's true, I glossed over this, yeah this is slightly different I guess you're right. We'll have to see if we can add a feature or if it needs to be a new primitive.

**Reply by DaniGuardiola** [reply]
If this is done properly, it's probably a popover with a combobox + combobox list inside. A combobox list doesn't need to necessarily be in a popover separate from the combobox (input) itself. Here's how you can achieve it with Ariakit: https://ariakit.org/examples/menu-combobox

It might be confusing to see `Menu` used there, but here's what Diego Haz (maintainer) told me: "Menu is MenuList + Popover, but it loses the menu semantics when you pass composite={false} to it".

Here's our component built with Ariakit (using this pattern): https://atlas-storybook.guide.co/?path=/story/components-menu--searchable

I guess once Combobox is ready, you could achieve something like this with Radix by composing it with a popover.

**Reply by benoitgrelard** [reply]
Not in the Linear case though, as it does hold a "value".

**Reply by DaniGuardiola** [reply]
You're completely right. This is still achievable in Ariakit by composing a `Select` and `Combobox` though, although I haven't had the chance to build this in our company yet, so I'm not super-familiar with it: https://ariakit.org/examples/select-combobox

**Reply by nandorojo** [reply]
https://user-images.githubusercontent.com/13172299/236477445-a8fc8847-0229-48d2-902e-5ae0e370780b.mp4



Spotify is a nice example of a dropdown menu with a search.

**Comment by johnhforrest** [comment]
I just wanted to chime in and say we'd love a component like this as well. We ended up building our own Combobox component and then putting it inside of a radix DropdownMenu primitive. It mostly "just works" but there are definitely some rough edges around focus management. Ideally I'd love to see this as some sort of FilterableDropdownMenu primitive. Here's some screenshots: one of a top level menu and one within a submenu.

<img width="232" alt="image" src="https://user-images.githubusercontent.com/402949/168385131-9a4b0f08-1787-4227-a8bb-844e790563db.png">

<img width="474" alt="image" src="https://user-images.githubusercontent.com/402949/168385000-d711e0b2-4a49-4510-acf9-2b09e1302d84.png">

**Reply by BrianHung** [reply]
How do you prevent focus from escaping the input box on dropdown menu item hover?

**Reply by misha-erm** [reply]
Is there any chance you can share the solution with the world?))

**Comment by Cakeday** [comment]
Just wanted to mention that adding `onKeyDown={(e: React.KeyboardEvent<HTMLInputElement>) => e.stopPropagation()}` to the input element, per @jjenzz's comment above, worked for me!

**Reply by mostafa-sakhiri** [reply]
Hello, this solution works for me too, but once I select an item, each time I change the input value, the input loses focus. Any idea about what causes this issue ?

**Reply by omtita-everag** [reply]
From the source code, It seems like the issue with the typeahead function of the component conflicting with the input's event.

**Comment by nawresncib12** [comment]
I tried something similar and it works fine but the input does not detect the spacebar onChange , it detects is onKeyDown but it's not being written and added to its value (not detected onChange) , this does not happen when the input is outside the dropdown trigger

**Comment by misha-erm** [comment]
for those who might be still looking on how to implement this pattern: [cmdk](https://github.com/pacocoursey/cmdk) plays pretty good with other radix primitives

https://codesandbox.io/p/sandbox/radix-dropdown-combobox-1x6e3k?file=%2Fsrc%2FApp.css%3A5%2C7-5%2C12

**Reply by misha-erm** [reply]
it has one known issue: autofocus doesn't really work when using mouse. But I was able to work around this with some focus locking (re-focus input onBlur if bluring back to submenu trigger with open menu)

**Comment by eldyvoon** [comment]
I'm waiting while using react-select first.

**Comment by klingat** [comment]
+1 also waiting on this!

**Comment by V-iktor** [comment]
👍

**Comment by florinkrasniqi** [comment]
+1 waiting on this!

Until then, something like this can do the job; Except for escaping input focus with keyboard navigation :| 

```
<input
  type="text"
  className="w-full border-none py-2.5 pl-9 pr-3 text-xs font-medium leading-6 placeholder:text-grey-300 focus:ring-0"
  placeholder="Search..."
  disabled={isLoading}
  value={search}
  onChange={(e) => setSearch(e.target.value)}
  onKeyDown={(e: React.KeyboardEvent<HTMLInputElement>) =>
    e.stopPropagation()
  }
/>
```

**Reply by mohamed-moubarak** [reply]
Try adding the following condition
```js
onKeyDown={(e: React.KeyboardEvent<HTMLInputElement>) =>
    if (e.key !== 'Escape') {
       e.stopPropagation();
     }
  }

```

**Comment by ankormoreankor** [comment]
Hi guys!
We also need this feature in our platform.
I really like your lib, it covers most of our cases and it works very well 💛
Currently, I'm going to try this approach - https://ariakit.org/examples/combobox-radix-select - but seems like I have to add the whole new lib in our project

**Reply by ankormoreankor** [reply]
Well, I've tried to implement this approach to DropdownMenu, and it didn't work for me: once I type a symbol, the popup closes, and autofocus is not working either(

**Reply by ankormoreankor** [reply]
Unfortunately, none of the approaches from this thread worked for me🫤

**Comment by eldyvoon** [comment]
Just use react-select

**Reply by ankormoreankor** [reply]
I'm thinking of using Downshift to write my own solution
As I understand, react-select **is not** a headless component, correct?

**Comment by rutpshah** [comment]
If anyone still searching for answer and want to stick with Radix UI then Use `https://www.radix-ui.com/primitives/docs/components/menubar`. Use menu items as dropdown items and change flex direciton as column. Add additional css modifications to make it look like dropdown.

Use `side` prop with value as `right` in Menubar content.

Push all this inside a popover along with search box.

**Comment by eszthoff** [comment]
Hello. Thanks for the enlightening discussion here. I have a different use case, where I want to have an input field inside a dropdown, no for filtering, but to add an new option. I think that's should be a legitimate usecase here, isn't it? 

I managed to add the input into the dropdown and make sure focus stays there while typing. However, I'm struggling to make both keyboard and mouse interaction work.

Keyboard navigation works with the following structure, but when the input it clicked, the dropdown is dismissed.
```
        <DropdownMenu.Root>
            <DropdownMenu.Trigger>Apply tags</DropdownMenu.Trigger>
            <DropdownMenu.Portal>
                <DropdownMenu.Content>
                    <DropdownMenu.Item asChild>
                        <input
                            type="text"
                            tabIndex={0}
                            onKeyDown={(e) => {
                                e.stopPropagation();
                            }}
                            placeholder="Create new..."
                            onChange={(e) => {
                                console.log('input change with e:', e);
                            }}
                            onClick={(e) => e.stopPropagation()}
                            onSelect={(e) => e.preventDefault()}
                        />
                    </DropdownMenu.Item>
                </DropdownMenu.Content>
            </DropdownMenu.Portal>
        </DropdownMenu.Root>
```

In the following example mouse interaction is as expected, but I cannot get focus on the input with the keyboard at all. 
```
        <DropdownMenu.Root>
            <DropdownMenu.Trigger>Apply tags</DropdownMenu.Trigger>
            <DropdownMenu.Portal>
                <DropdownMenu.Content>
                        <input
                            type="text"
                            tabIndex={0}
                            onKeyDown={(e) => {
                                e.stopPropagation();
                            }}
                            placeholder="Create new..."
                            onChange={(e) => {
                                console.log('input change with e:', e);
                            }}
                            onClick={(e) => e.stopPropagation()}
                            onSelect={(e) => e.preventDefault()}
                        />
                </DropdownMenu.Content>
            </DropdownMenu.Portal>
        </DropdownMenu.Root>
```

Any advise?

**Reply by eszthoff** [reply]
For anyone else looking here... It was a mistake in the `onClick` handler. It should call `e.PreventDefault()` not `e.stopProoagation()`. With that change, our first example (`<DropdownMenu.Item asChild><Input/></DropdownMenu.Item asChild>` works as expected. 

An other thing I didn't see in the notes, is that we need to forwardRef in case we use our own component here. E.g. we want to set focus on the input field when the popup opens via ref, we had to sure the ref is forwarded correctly.
